### PR TITLE
Offsite aware redirect 

### PR DESCRIPTION
**Pull request by aliowka**
Here is my attempt to solve this issue https://github.com/scrapy/scrapy/issues/15.

### CODE REVIEW

File: scrapy/contrib/downloadermiddleware/redirect_offsite.py [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+        return self.offsite_filtered(redirected, spider)
+
+    def offsite_filtered(self, redirected, spider):
+
+        allowed_domains = getattr(spider, 'allowed_domains', None)
+
+        if not allowed_domains:
+            return redirected
+
+        if isinstance(redirected, Request):
```
Comment (stav) [inline scrapy/contrib/downloadermiddleware/redirect_offsite.py]: This method is called from `process_response`, how can the redirected ever be a `Request` here?

Comment (aliowka) [inline scrapy/contrib/downloadermiddleware/redirect_offsite.py]: From the documentation: 
```process_response()``` should either: return a ```Response``` object, return a ```Request``` object or raise a ```IgnoreRequest``` exception.
(http://doc.scrapy.org/en/latest/topics/downloader-middleware.html#scrapy.contrib.downloadermiddleware.DownloaderMiddleware.process_response)
So there is no problem to return a ```Request``` from ```process_response```.
In ```RedirectMiddleware.process_response``` method htat is used here there is a number of cases when the response contains a redirection, so a new ```Request``` is returned that is pointing to the new destination.

Comment (stav) [inline scrapy/contrib/downloadermiddleware/redirect_offsite.py]: Oh yeah, sorry I was using downloader middleware.  Spider can return another Request, of course..

Comment (stav) [inline scrapy/contrib/downloadermiddleware/redirect_offsite.py]: Ok, you are sub-classing `RedirectMiddleware` which is doing the redirecting, gotcha.

**Comment by Gallaecio** [comment]
It’s been a long time, and judging by #2241 this topic requires some further discussion. @aliowka would you be OK with closing this pull request, or would you rather continue working on it?

**Comment by aliowka** [comment]
Thanks for pinging me @Gallaecio , indeed it's been a while since this PR.

I would like to take time and go through the discussion #2241 in order to fully understand the concerns about (not) merging this PR, and then I'll be able to decide whether I'm going to fix it or to close.
This might take several days. 
Thanks for understanding..

**Pull request closed without merging**